IQSS/dataverse
# 9231 Custom HTTP headers in OAI requests.

**What this PR does / why we need it**:

This is a feature request from our Elseveir partners. In their case, the remote OAI server requires an extra auth. token, supplied in an http header with every OAI request (something along the lines of `x-api-key: xxx-yyy-zzz`) in order to serve some content not available to other clients. See the linked issue. So this pr adds the ability to add this header to the harvesting client configuration, both in the GUI and the api.

The changes are fairly straightforward.  The only moderately tricky part was it required a PR into the gdcc.xoa library project as well (gdcc/xoai/pull/119; merged). 

**Which issue(s) this PR closes**:

Closes #9231

**Special notes for your reviewer**:

Please note that the branch in its current state builds with the _snapshot_ (`-5.0.0-SNAPSHOT`) versions of the XOAI library jars. These jars are automatically built on the xoai side from the 5.0 branch in its present state, i.e., with the changes that were necessary to accommodate this feature, that were made in gdcc/xoai/pull/119 that has been merged recently. While the PR should be fully ready for review and testing, it will be better to switch these jars to a stable minor release version in the pom file before the PR is merged. (The jars will have the exact same code in them, providing the exact same functionality, but in fixed version jars guaranteed to be available on mvn). I'll do this once that release - `5.0.0-RC3` - is tagged, this should happen very shortly on the gdcc.xoai side.  

Regarding the Harvesting Clients UI: I initially assumed we would bury this extra header setting on the last page of the create/edit dialogue, under something like "advanced settings" - on account of how very few admins will have a use for this setting. I realized however that it needed to be right there on the first, "Step 1." page - because the remote server may require this token even to be willing to answer the basic "ListSets" and "ListMetadataFormats" calls, that must happen during the setting up of the client, in "Step 2". The field is clearly marked as "optional"; I don't expect it to confuse anyone too much.

**Suggestions on how to test this**:

All we need to test is that an admin can add some extra header to the configuration, and the Harvester will dutifully add it to every request sent for this client. We don't need to test any scenario where this token actually *does* something. So a straightforward way to test this would be to: 
- create a client harvesting some small set from another server we can easily reconfigure - for example, my "controlTestSet" on demo.dataverse.org;
- add some arbitrary header to the configuration; say `x-oai-api-key: 111-222-333`, which is roughly what it'll look like in the Elseveir use case;
- add the header above to the access log format on the oai server (demo) side; either on the Apache or Payara level. Confirm that the header is in fact there. 

That, and some basic check that the PR isn't breaking anything obvious for existing clients that don't have this custom header feature configured.

**ONE MORE THING:** The functionality in the PR is ready to be tested. But a minor thing in the pom file still needs to be changed before we merge it - the application in this branch is being built with the dev. version of the XOAI libraries; I'm waiting for it to be tagged as the next point release - "5.0.0-RC3" on the XOAI side, and once that's done, I'll need to change the pom file accordingly. I'm hoping that this will happen very shortly, but please confirm with me before merging - thank you!

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

A minor change to the Harvesting Clients admin GUI, as described above: 

![Screen Shot 2023-01-23 at 11 49 05 AM](https://user-images.githubusercontent.com/737558/214099583-ed2559f7-9cc2-4e0f-8b2c-eb8f2c4a9a85.png)

**Is there a release notes update needed for this change?**:

**Additional documentation**:
